ll2_global_planner/src/ll2_global_planner_nodelet.cpp

```diff
@@ -44,10 +49,178 @@ void Ll2GlobalPlannerNl::loadParams()
+    int wp_length = centerline.size() - 1;
+
+    // Loop over each centerline point
+    for (int i = 0; i <= wp_length; i++)
+    {
+      auto point = centerline[i];
+
+      autoware_msgs::Waypoint new_wp;
+      new_wp.pose.pose.position.x = point.x();
+      new_wp.pose.pose.position.y = point.y();
+      new_wp.pose.pose.position.z = point.z();
+      new_wp.twist.twist.linear.x = speed_limit.speedLimit.value(); // m/s
+
+      int steering_state = autoware_msgs::WaypointState::STR_STRAIGHT;
+      if (turn_direction.compare("right") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_RIGHT;
+      }
+      if (turn_direction.compare("left") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_LEFT;
+      }
+      new_wp.wpstate.steering_state = steering_state;
+
+      waypoints.push_back(new_wp);
+    }
+  }
+
+  int wp_id = 0;
+  bool first_point = true;
+  float smallest_goal_dist = std::numeric_limits<float>::infinity();
+  int smallest_goal_wp_id = 0;
+  autoware_msgs::Waypoint* prev_waypoint = nullptr;
+  std::vector<autoware_msgs::Waypoint> processed_waypoints;
+
+  // Remove duplicates and set waypoint orientations
+  for (autoware_msgs::Waypoint& waypoint : waypoints)
+  {
+    geometry_msgs::Pose& pose = waypoint.pose.pose;
+    geometry_msgs::Pose& prev_pose = prev_waypoint->pose.pose;
+
+    if (!first_point)
+    {
+      // Check for duplicate points
+      float dist = std::hypot(pose.position.x - prev_pose.position.x, pose.position.y - prev_pose.position.y);
+      if (dist < duplicate_dist_threshold)
+      {
+        continue;
+      }
+
+      // Set orientation of the previous point, aiming at current point
+      float yaw = std::atan2(pose.position.y - prev_pose.position.y, pose.position.x - prev_pose.position.x);
+
+      tf2::Quaternion orientation;
+      orientation.setRPY(0, 0, yaw);
+      tf2::convert(orientation, prev_pose.orientation);
+    }
+    else
+    {
+      first_point = false;
+    }
+
+    // Calculate distance to goal point
+    float goal_dist = std::hypot(std::hypot(goal_point.x - pose.position.x, goal_point.y - pose.position.y), goal_point.z - pose.position.z);
+    if (goal_dist < smallest_goal_dist)
+    {
+      smallest_goal_dist = goal_dist;
+      smallest_goal_wp_id = wp_id;
+    }
+
+    waypoint.gid = wp_id;
+    processed_waypoints.push_back(waypoint);
+    prev_waypoint = &processed_waypoints.back();
+    wp_id++;
+  }
+
+  // Set orientation of the last waypoint
+  processed_waypoints.back().pose.pose.orientation = processed_waypoints.end()[-2].pose.pose.orientation;
+
+  autoware_msgs::Lane lane_msg;
+  // Trim waypoints to stop at goal point
+  std::move(processed_waypoints.begin(), processed_waypoints.begin() + smallest_goal_wp_id, std::back_inserter(lane_msg.waypoints));
+
+  lane_msg.header.stamp = ros::Time(0);
+  lane_msg.header.frame_id = "map";
+  lane_msg.is_blocked = false;
```

Comment:
Perhpas this flag needs to have three states: unchecked, clear and blocked?

Reply:
I'm honestly not sure if anything downstream even uses this flag, but if it does, it's likely the responsibility of some perception node to update this field based on objects in the path.